spec/ics-004-channel-and-packet-semantics/README.md

```diff
@@ -599,41 +604,75 @@ function recvPacket(
+    } else {
+      // for unordered channels we must set the receipt so it can be verified on the other side
+      abortTransactionUnless(provableStore.get(packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence) === null))
+      provableStore.set(
+        packetReceiptPath(packet.destPort, packet.destChannel, packet.sequence),
+        ""
+      )
+    }
 
-    // log that a packet has been received & acknowledged
+    // log that a packet has been received
     emitLogEntry("recvPacket", {sequence: packet.sequence, timeoutHeight: packet.timeoutHeight,
-                                timeoutTimestamp: packet.timeoutTimestamp, data: packet.data, acknowledgement})
+                                timeoutTimestamp: packet.timeoutTimestamp, data: packet.data})
 
     // return transparent packet
     return packet
 }
 ```
 
-#### Acknowledgements
+#### Writing acknowledgements
+
+The `writeAcknowledgement` function is called by a module in order to write data which resulted from processing an IBC packet that the sending chain can then verify, a sort of "execution receipt" or "RPC call response".
+
+Calling modules MUST execute application logic atomically in conjunction with calling `writeAcknowledgement`.
+
+This is an asynchronous acknowledgement, the contents of which do not need to be determined when the packet is received, only when processing is complete. In the synchronous case, `writeAcknowledgement` can be called in the same transaction (atomically) with `recvPacket`.
+
+Acknowledging packets is not required; however, if an ordered channel uses acknowledgements, either all or no packets must be acknowledged (since the acknowledgements are processed in order).
+
+`writeAcknowledgement` *does not* check if the packet being acknowledged was actually received, because this would result in proofs being verified twice for acknowledged packets. This aspect of correctness is the responsibility of the calling module.
+The calling module MUST only call `writeAcknowledgement` with a packet previously received from `recvPacket`.
+
+The IBC handler performs the following steps in order:
+
+- Checks that an acknowledgement for this packet has not yet been written
+- Sets the opaque acknowledgement value at a store path unique to the packet
+
+```typescript
+function writeAcknowledgement(
+  packet: Packet,
+  acknowledgement: bytes): Packet {
+
+    // cannot already have written the acknowledgement
+    abortTransactionUnless(provableStore.get(packetAcknowledgementPath(packet.destPort, packet.destChannel, packet.sequence) === null))
+
+    // write the acknowledgement
+    provableStore.set(
+      packetAcknowledgementPath(packet.destPort, packet.destChannel, packet.sequence),
+      hash(acknowledgement)
+    )
+
+    // log that a packet has been acknowledged
+    emitLogEntry("writeAcknowledgement", {sequence: packet.sequence, timeoutHeight: packet.timeoutHeight,
```

Comment:
shouldn't it emit the channel end port ID and channel ID the ack was written to (ie packet.destPort, packet.destChannel)?

Reply:
Aye, added.